arangod/Cluster/RebootTracker.cpp

```diff
@@ -66,296 +95,116 @@ RebootTracker::RebootTracker(RebootTracker::SchedulerPointer scheduler)
-        std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-        callbacks) {
-  TRI_ASSERT(!callbacks.empty());
-  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
-                          [](auto it) { return it == nullptr; }));
-  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
-                          [](auto it) { return it->empty(); }));
-
-  return [callbacks = std::move(callbacks)]() {
-    LOG_TOPIC("80dfe", DEBUG, Logger::CLUSTER)
-        << "Executing scheduled reboot callbacks";
-    TRI_ASSERT(!callbacks.empty());
-    for (auto const& callbacksPtr : callbacks) {
-      TRI_ASSERT(callbacksPtr != nullptr);
-      TRI_ASSERT(!callbacksPtr->empty());
-      for (auto const& it : *callbacksPtr) {
-        auto const& cb = it.second.callback;
-        auto const& descr = it.second.description;
-        LOG_TOPIC("afdfd", DEBUG, Logger::CLUSTER)
-            << "Executing callback " << it.second.description;
-        try {
-          cb();
-        } catch (arangodb::basics::Exception const& ex) {
-          LOG_TOPIC("88a63", INFO, Logger::CLUSTER)
-              << "Failed to execute reboot callback: " << descr << ": "
-              << "[" << ex.code() << "] " << ex.what();
-        } catch (std::exception const& ex) {
-          LOG_TOPIC("3d935", INFO, Logger::CLUSTER)
-              << "Failed to execute reboot callback: " << descr << ": "
-              << ex.what();
-        } catch (...) {
-          LOG_TOPIC("f7427", INFO, Logger::CLUSTER)
-              << "Failed to execute reboot callback: " << descr << ": "
-              << "Unknown error.";
-        }
-      }
-    }
+  auto schedule = [&](auto&& batch) noexcept {
+    _scheduler->queue(RequestLane::CLUSTER_INTERNAL,
+                      [batch = std::move(batch)]() mutable noexcept {
+                        for (auto& callbacks : batch) {
+                          safeInvokes(callbacks);
+                        }
+                      });
   };
-}
-
-void RebootTracker::queueCallbacks(
-    std::vector<
-        std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>>
-        callbacks) {
-  if (callbacks.empty()) {
+  auto& reboots = it->second;
+  TRI_ASSERT(!reboots.empty());
+  if (to > reboots.crbegin()->first) {
+    schedule(std::move(it->second));
+    _callbacks.erase(it);
     return;
   }
-
-  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
-                          [](auto it) { return it == nullptr; }));
-  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
-                          [](auto it) { return it->empty(); }));
-
-  auto cb = createSchedulerCallback(std::move(callbacks));
-  _scheduler->queue(RequestLane::CLUSTER_INTERNAL, std::move(cb));
+  std::vector<RebootIds::node_type> callbacks;
+  for (auto rbIt = reboots.begin(); rbIt->first < to;) {
+    TRI_ASSERT(rbIt != reboots.end());
+    callbacks.push_back(reboots.extract(rbIt++));
+  }
```

Comment:
Nitpicking, but wouldn't
```suggestion
  for (auto rbIt = reboots.begin(); rbIt->first < to; ++rbIt) {
    TRI_ASSERT(rbIt != reboots.end());
    callbacks.push_back(reboots.extract(rbIt));
  }
```
be more idiomatic?

Reply:
Iterator will be invalid after extract

Reply:
Oh, true.